### PR TITLE
MAM-3499-performance-image-transforming-decoding-and-caching

### PR DESCRIPTION
**Pull request by bnolens**
https://linear.app/theblvd/issue/MAM-3499/performance-image-transforming-decoding-and-caching

### CODE REVIEW

File: Mammoth/Views/Cells/PostCardCell/PostCardProfilePic.swift [hunk 1]
```diff
@@ -152,19 +166,23 @@ extension PostCardProfilePic {
...
+                    placeholderImage: nil,
+                    context: [.imageTransformer: PostCardProfilePic.transformer],
+                    progress: nil
+                ) { [weak self] image, error, _, finished in
+                        guard let self else { return }
+                        if let image {
+                            self.user?.decodedProfilePic = image
+                        }
+                }
+            }
```
Comment (rileyhCode) [inline Mammoth/Views/Cells/PostCardCell/PostCardProfilePic.swift:183]: This code/pattern is repeated a few times. Is it possible to encapsulate it?

**Review by rileyhCode — CHANGES_REQUESTED** [review]
A few questions/comments in the review.

File: Mammoth/Screens/DetailScreen/DetailViewModel.swift [hunk 2]
```diff
@@ -249,16 +249,26 @@ extension DetailViewModel {
...
                     parents?.forEach({
                         $0.preloadQuotePost()
-                        $0.preloadImages()
                     })
+                    
+                    PostCardModel.imageDecodeQueue.async {
+                        parents?.forEach({
+                            $0.preloadImages()
+                        })
+                    }
```
[the remark this answers is not in the document]
Comment (bnolens) [reply Mammoth/Screens/DetailScreen/DetailViewModel.swift:261]: I like that!

File: Mammoth/Models/PostCardModel.swift [hunk 3]
```diff
@@ -734,6 +752,9 @@ extension PostCardModel {
         self.preloadedImageURLs.forEach({
             SDImageCache.shared.removeImageFromMemory(forKey: $0)
         })
+        
+        self.decodedImages = [:]
+        self.user?.clearCache()
```
[the remark this answers is not in the document]
Comment (bnolens) [reply Mammoth/Models/PostCardModel.swift:757]: Good point!

File: Mammoth/Views/Cells/PostCardCell/PostCardProfilePic.swift (hunk 1 above)
Comment (bnolens) [reply Mammoth/Views/Cells/PostCardCell/PostCardProfilePic.swift:183]: Not sure I see how to encapsulate this and really improve the code. 

Here's a function encapsulating the logic, but to me it make the code less readable. 

Screenshot 2023-12-15 at 12 05 38

Comment (bnolens) [reply Mammoth/Views/Cells/PostCardCell/PostCardProfilePic.swift:183]: Or as an UIImageView extension, which I like a little better already. 

Screenshot 2023-12-15 at 12 17 35

Comment (bnolens) [reply Mammoth/Views/Cells/PostCardCell/PostCardProfilePic.swift:183]: Implementation: 
Screenshot 2023-12-15 at 12 18 24

Comment (bnolens) [reply Mammoth/Views/Cells/PostCardCell/PostCardProfilePic.swift:173]: I ended up using the second version and pushed a commit with these changes

**Review by rileyhCode — APPROVED** [review]
LGTM

**Pull request merged**